r2-core/src/main/java/com/linkedin/r2/transport/http/client/EvictingCircularBuffer.java

```diff
@@ -0,0 +1,245 @@
+  private final AtomicInteger _writerPosition = new AtomicInteger();
+  private final Clock _clock;
+
+
+  public EvictingCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit)
+  {
+    this(capacity, ttl, ttlUnit, SystemClock.instance());
+  }
+
+  EvictingCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit, Clock clock)
+  {
+    setCapacity(capacity);
+    setTtl(ttl, ttlUnit);
+    _clock = clock;
+  }
+
+  /**
+   * Adds the supplied Callback to the internal circular buffer. If the buffer is full, the oldest Callback in the buffer
+   * will be overwritten. Calls to put always succeed, and there is no guarantee that Callbacks submitted through the put()
+   * method will be subsequently returned by the get() method.
+   *
+   * @param toAdd Callback that is to be possibly returned by later calls to get()
+   */
+  public void put(Callback<None> toAdd)
+  {
+    int writerPosition = getAndBumpWriterPosition();
+    ReentrantReadWriteLock thisLock = _elementLocks.get(writerPosition);
+    thisLock.writeLock().lock();
+    try
+    {
+      _callbacks.set(writerPosition, toAdd);
+      _ttlBuffer.set(writerPosition, Instant.ofEpochMilli(_clock.currentTimeMillis()));
+    } finally {
+      thisLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Returns a Callback previously stored in the circular buffer through the put() method. Callbacks are generally returned
+   * in the order they were received, but in cases of get() throughput in excess of put() throughput, previously returned
+   * Callbacks will be sent. Oldest age of returned Callbacks is configurable through the ttl param in the constructor.
+   *
+   * Calls to get() will always succeed as long as calls to put() continue at a cadence within the ttl duration. When write
+   * throughput has dropped to zero, get() will eventually throw NoSuchElementException once the circular buffer has become
+   * fully pruned through expired ttl.
+   * @return Callback
+   * @throws NoSuchElementException if internal circular buffer is empty
+   */
+  public Callback<None> get()
+  {
+    int thisReaderPosition = getAndBumpReaderPosition();
+    ReentrantReadWriteLock thisLock = _elementLocks.get(thisReaderPosition);
+
+    thisLock.readLock().lock();
+    Callback<None> callback;
+    Instant ttl;
+    try {
+      callback = _callbacks.get(thisReaderPosition);
+      ttl = _ttlBuffer.get(thisReaderPosition);
+    }
+    finally
+    {
+      thisLock.readLock().unlock();
+    }
+
+    if (callback == null)
+    {
+      if (thisReaderPosition == 0)
+      {
+        throw new NoSuchElementException("Buffer is empty");
+      }
+      else
+      {
+        return get();
+      }
+    }
+    else if (Duration.between(ttl, Instant.ofEpochMilli(_clock.currentTimeMillis())).compareTo(_ttl) > 0)
+    {
+      if (thisReaderPosition != 0 || this.size() == 1)
+      {
+        thisLock.writeLock().lock();
+        try {
+          if (callback == _callbacks.get(thisReaderPosition))
```

Comment:
What's the purpose of this check?

Reply:
Added comment above the check in the code. Since we release the read lock prior to this point, it's possible a write could have changed the value from underneath us. This check is to ensure that, once we have the write lock, the value is still the same as what we thought it was.